src/deluge/util/string.cpp

```
@@ -0,0 +1,83 @@
+#include "util/string.h"
+#include "printf.h"
+#include "util/lookuptables/lookuptables.h"
+#include "util/try.h"
+#include <array>
+#include <cmath>
+#include <cstdint>
+#include <expected>
+#include <string>
+#include <system_error>
+
+namespace deluge {
+std::expected<char*, std::errc> to_chars(char* first, char* last, float value, int precision) {
+	if (first >= last) {
+		return std::unexpected{std::errc::no_buffer_space};
+	}
+
+	size_t buffer_size = last - first;
+
+	std::array<char, 16> format;
+	snprintf_(format.data(), format.size(), "%%.%df", precision);
+
+	int written = snprintf_(first, buffer_size, format.data(), value);
+	if (written < 0 || static_cast<size_t>(written) >= buffer_size) {
+		std::unexpected{std::errc::no_buffer_space};
+	}
+	return first + written;
+}
+} // namespace deluge
+
+namespace deluge::string {
+std::string fromInt(int32_t number, size_t min_num_digits) {
+	// Convert number to string
+	std::string result = std::to_string(number);
+
+	// Add leading zeros if needed
+	if (result[0] != '-' && result.size() < min_num_digits) {
+		result.insert(result.begin(), min_num_digits - result.size(), '0');
+	}
+	else if (result[0] == '-' && result.size() - 1 < min_num_digits) {
+		result.insert(result.begin() + 1, min_num_digits - (result.size() - 1), '0');
+	}
+
+	return result;
+}
+
+std::string fromFloat(float number, int32_t precision) {
+	std::array<char, 64> buffer;
+	char* ptr = D_TRY_CATCH(deluge::to_chars(buffer.data(), buffer.data() + buffer.size(), number, precision), error, {
```

Should this return an error of some sort instead of an empty string?

---

Added comment to clarify reason for return, per discord discussion